vortex-expr/src/forms/cnf.rs

```diff
@@ -0,0 +1,169 @@
+/// parameters:
+///
+/// ```
+/// use vortex_expr::{not, col, or, and, gt_eq, lit, not_eq, lt, eq};
+/// use vortex_expr::forms::cnf::cnf;
+/// use itertools::Itertools;
+///
+/// assert_eq!(
+///     cnf(
+///         or(
+///             and(
+///                 gt_eq(col("earnings"), lit(50_000)),
+///                 not_eq(col("role"), lit("Manager"))
+///             ),
+///             col("special_flag")
+///         ),
+///     ).unwrap(),
+///     vec![
+///         vec![
+///             gt_eq(col("earnings"), lit(50_000)),
+///             col("special_flag")
+///         ],
+///         vec![
+///             not_eq(col("role"), lit("Manager")),
+///             col("special_flag")
+///         ]
+///     ]
+/// );
+/// ```
+///
+/// ```
+/// use vortex_expr::{not, col, or, and, gt_eq, lit, not_eq, lt, eq};
+/// use vortex_expr::forms::cnf::cnf;
+/// use itertools::Itertools;
+///
+/// assert_eq!(
+///     cnf(
+///         or(
+///             or(
+///                 and(
+///                     gt_eq(col("earnings"), lit(50_000)),
+///                     not_eq(col("role"), lit("Manager"))
+///                 ),
+///                 col("special_flag")
+///             ),
+///             and(
+///                 lt(col("tenure"), lit(5)),
+///                 eq(col("role"), lit("Engineer"))
+///             ),
+///         )
+///     ).unwrap(),
+///     vec![
+///         vec![
+///             gt_eq(col("earnings"), lit(50_000)),
+///             col("special_flag"),
+///             lt(col("tenure"), lit(5)),
+///         ],
+///         vec![
+///             gt_eq(col("earnings"), lit(50_000)),
+///             col("special_flag"),
+///             eq(col("role"), lit("Engineer")),
+///         ],
+///         vec![
+///             not_eq(col("role"), lit("Manager")),
+///             col("special_flag"),
+///             lt(col("tenure"), lit(5)),
+///         ],
+///         vec![
+///             not_eq(col("role"), lit("Manager")),
+///             col("special_flag"),
+///             eq(col("role"), lit("Engineer")),
+///         ],
+///     ]
+/// );
+/// ```
+///
+pub fn cnf(expr: ExprRef) -> VortexResult<Vec<Vec<ExprRef>>> {
+    let nnf = nnf(expr)?;
+
+    let mut visitor = CNFVisitor::default();
+    nnf.accept(&mut visitor)?;
+    Ok(visitor.finish())
+}
+
+#[derive(Default)]
+struct CNFVisitor {
+    conjuncts_of_disjuncts: Vec<Vec<ExprRef>>,
+}
+
+impl CNFVisitor {
+    fn finish(self) -> Vec<Vec<ExprRef>> {
+        self.conjuncts_of_disjuncts
+    }
+}
+
+impl NodeVisitor<'_> for CNFVisitor {
+    type NodeTy = ExprRef;
+
+    fn visit_down(&mut self, node: &ExprRef) -> VortexResult<TraversalOrder> {
+        if let Some(binary_expr) = node.as_any().downcast_ref::<BinaryExpr>() {
+            match binary_expr.op() {
+                Operator::And => return Ok(TraversalOrder::Continue),
+                Operator::Or => {
```

Comment:
I am not sure why we have the disjuncts at all?

Reply:
I just don’t see why we want this. how will we use these to execute sections of a query in a layout

Reply:
Are you suggesting that if we see an OR we should bail out of the CNF construction entirely because it is unlikely to bring a benefit?

I agree that disjuncts basically kill our ability to reason in "isolation".

Reply:
Yep, but I am not that fussed either way